src/h3lib/lib/edge.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2022 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file  edge.c
+ * @brief Edge functions for manipulating (undirected) edge indexes.
+ */
+
+#include <inttypes.h>
+#include <stdbool.h>
+
+#include "algos.h"
+#include "constants.h"
+#include "coordijk.h"
+#include "h3Assert.h"
+#include "h3Index.h"
+#include "latLng.h"
+#include "vertex.h"
+
+/**
+ * Wrap the error code from a directed edge function and present
+ * undirected edge errors instead.
+ */
+H3Error wrapDirectedEdgeError(H3Error err) {
+    if (err == E_DIR_EDGE_INVALID) {
+        return E_UNDIR_EDGE_INVALID;
+    }
+    return err;
+}
+
+/**
+ * Allows for operations on an edge index as if it were a directed edge
+ * from the edge owner to the neighboring cell.
+ * @param edge Undirected edge index
+ * @return H3Index Directed edge index
+ */
+H3Index edgeAsDirectedEdge(H3Index edge) {
+    // Do not make indexes that are not edges look "valid".
+    if (H3_GET_MODE(edge) == H3_EDGE_MODE) {
+        H3_SET_MODE(edge, H3_DIRECTEDEDGE_MODE);
+    }
+    return edge;
+}
+
+/**
+ * Returns an edge H3 index based on the provided neighboring cells
+ * @param cell1 An H3 hexagon index
+ * @param cell2 A neighboring H3 hexagon index
+ * @param out Output: the edge H3Index
+ */
+H3Error H3_EXPORT(cellsToEdge)(H3Index cell1, H3Index cell2, H3Index *out) {
+    bool cell1IsOrigin = cell1 < cell2;
+    H3Index origin = cell1IsOrigin ? cell1 : cell2;
+    H3Index dest = cell1IsOrigin ? cell2 : cell1;
+    H3Error edgeErr = H3_EXPORT(cellsToDirectedEdge)(origin, dest, out);
+    if (!edgeErr) {
+        H3_SET_MODE(*out, H3_EDGE_MODE);
+        return edgeErr;
+    } else {
+        return edgeErr;
+    }
+}
```

Comment:
The boolean check based on the numeric value of the indexes will make sense most of the time, but will probably cause unexpected weirdness along base cell boundaries. I don't have a better idea here, though.

Also, perhaps `cellsToDirectedEdge` should be macro-ified and re-used in both, without needing to set the mode bit twice?

Reply:
Could you describe the weirdness you are anticipating? I am not sure what would be odd. I expect the cell with the lower base cell number would always be chosen along the boundary with another base cell.

Reply:
So here's a crude diagram showing the underlying directed edges for the 12 undirected edges amongst a set of "sibling" cells.

![undirected_edge_inner_directions](https://user-images.githubusercontent.com/765551/213018377-6dd8615f-441c-4ac0-9d91-3e205ed15110.png)

There's an asymmetry in the number of edges that are the origin or destination for any given cell, with the center cells having 6 origins and no destinations, the 3 "earlier" siblings after that having 2 origins and 1 destination, and the 3 "later" siblings having no origins and 3 destinations.

That in itself is a little odd, but not terrible in that it repeats. Cells that aren't "siblings" will be harder to predict for users, but make sense in a fashion as long as they have the same base cell, as the orientation of the edge will be determined by the tree structure of their parent cells' border, or grandparent cells' border, etc.

But once you get to the base cells, they are simply hardwired on each icosahedron face and this pattern breaks and there's no spatial rhyme or reason to the selection.

When this was proposed I was assuming we'd do something like the local IJ coordinate system and choose the cell with the smaller absolute value of the magnitude of the IJ vector or something, and thought the holdup was just figuring out a performant way to do it.

This would have some spatial relationship, in that the vectors would tend to point "out from" the center of each icosahedron face and point into each pentagon, and hexagons along the border of the icosahedron face would, I believe, tend to point east in roughly the north pole region, west in the south pole region, and alternate east and west in the equatorial "band" of triangle faces, where the vector sum of all of these edges should be zero.

I had mentioned the "smallest integer index wins" approach when I did the first version of directed edges, but as a "that's terrible" deprecating statement that I couldn't figure out a better solution at the time.